### PR TITLE
Added Andes Message Component

### PR DESCRIPTION
**Pull request by lautimfernandez**
## Motivation
Adding Andes Message component to the lib.

## Description
Andes Message Component in XML way was added without test and no programmatic way.

## Affected component
Andes Message

## Frontify component link
https://company-161429.frontify.com/d/kxHCRixezmfK/n-a#/componentes/message

## Screenshots
Screenshot_1579808391

## Checks
Are you sure you followed all those steps? In such case, let's say with me "I solemnly declare that ...":
- [X] I have coded an example inside the Demo App,
- [ ] I have added proper tests,
- [X] I have modified the [Changelog](https://github.com/mercadolibre/...) file,
- [ ] I have updated GitHub wiki,
- [ ] I have the explicit approval of one or more members of the UX Team,
- [X] I have updated the version in gradle.properties file.

## Change type
This is easy, let us know what this code is about:
- [X] This code adds a new component
- [ ] This code includes improvements to an existent component
- [ ] This code improves or modifies ONLY the Demo App.

### CODE REVIEW

**Comment by furyGorgoryAdmin** [comment]
### Informative Menu <h2>  ### 

### Use any of these tags in a pull request description or leave a comment ### 
| Tag | Description |
|---|---|
| **[canary]**| Creates a build that can only be deployed with canary strategy. | 
| **[build]** |  Creates version or build from a branch that has been tagged as "no releasable". | 
| **[name:version-name]** |Specify the name of the build. For example: [build] [name:1.2.3-name] or [canary] [name:1.2.3-name].|
| **[major]** | To automatically increase the first position of the version. For example: [build] [major] or [canary] [major].|
| **[minor]** | To automatically increase the second position of the version. For example: [build] [minor] or [canary] [minor]. |
| **[patch]** | To automatically increase the third position of the version. For example: [build] [patch] or [canary] [patch].|  

[For more information about ci-cd](http://docs.furycloud.io/?lang=esp&section=businessassurance)
[For more information about semver](https://semver.org/#summary)

File: components/src/main/java/com/mercadolibre/android/andesui/message/AndesMessage.kt [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+
+    constructor(context: Context, attrs: AttributeSet?) : super(context) {
+        initAttrs(attrs)
+    }
+
+    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(context) {
+        initAttrs(attrs)
+    }
+
+    constructor(context: Context, andesMessageHierarchy: AndesMessageHierarchy, andesMessageState: AndesMessageState, body: String, title: String? = null, isDismissable: Boolean = IS_DISMISSABLE_DEFAULT) : super(context) {
```
Comment (jiurchuk) [inline components/src/main/java/com/mercadolibre/android/andesui/message/AndesMessage.kt]:
```suggestion
    constructor(context: Context, andesMessageHierarchy: AndesMessageHierarchy, andesMessageState: AndesMessageState, body: String, title: String? = TITLE_DEFAULT, isDismissable: Boolean = IS_DISMISSABLE_DEFAULT) : super(context) {
```

File: components/src/main/java/com/mercadolibre/android/andesui/message/AndesMessage.kt [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+            id = View.generateViewId()
+        }
+    }
+
+    /**
+     * Gets data from the config and sets to the text component of this button.
+     *
+     */
+    private fun setupTitleComponent() {
+        if(config.titleText == "" || config.titleText == null) {
```
Comment (jiurchuk) [inline components/src/main/java/com/mercadolibre/android/andesui/message/AndesMessage.kt]:
```suggestion
        if(config.titleText == null || config.titleText == "") {
```

File: components/src/main/java/com/mercadolibre/android/andesui/message/AndesMessage.kt [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+            titleComponent.visibility = View.GONE
+        }
+        else {
+            titleComponent.text = config.titleText
+            titleComponent.visibility = View.VISIBLE
```
Comment (jiurchuk) [inline components/src/main/java/com/mercadolibre/android/andesui/message/AndesMessage.kt]: Pongamos esta línea primero o última

File: components/src/main/java/com/mercadolibre/android/andesui/message/hierarchy/AndesMessageHierarchyInterface.kt [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    override fun iconBackgroundColor(context: Context, state: AndesMessageStateInterface) = state.secondaryColor(context)
+}
+
+internal object AndesQuietMessageHierarchy : AndesMessageHierarchyInterface() {
+    override fun backgroundColor(context: Context) = ContextCompat.getColor(context, R.color.andesui_message_quiet_bg) //TODO Check
+    override fun backgroundColor(context: Context, state: AndesMessageStateInterface) = backgroundColor(context)
+    override fun textColor(context: Context) = ContextCompat.getColor(context, R.color.andesui_message_quiet_text) //TODO Check
+    override fun dismissableIconColor(context: Context) = ContextCompat.getColor(context, R.color.andesui_message_quiet_dismissable)
+    override fun iconBackgroundColor(context: Context, state: AndesMessageStateInterface) = state.primaryColor(context)
+
```
Comment (jiurchuk) [inline components/src/main/java/com/mercadolibre/android/andesui/message/hierarchy/AndesMessageHierarchyInterface.kt]: Empty line

File: components/src/main/java/com/mercadolibre/android/andesui/utils/DrawableUtils.kt [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+import android.graphics.drawable.BitmapDrawable
+import android.graphics.drawable.shapes.OvalShape
+import android.graphics.drawable.ShapeDrawable
+import android.graphics.drawable.Drawable
+import android.graphics.drawable.LayerDrawable
+
+
+
+
+
```
Comment (jiurchuk) [inline components/src/main/java/com/mercadolibre/android/andesui/utils/DrawableUtils.kt]: Empty lines

File: components/src/main/java/com/mercadolibre/android/andesui/utils/DrawableUtils.kt [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+            .apply {
+                color?.let { setColorFilter(it, PorterDuff.Mode.SRC_IN) }
+            }
+}
+
+internal fun buildColoredCircularShapeWithIconDrawable(image: BitmapDrawable, context: Context, iconColor: Int? = null, shapeColor: Int? = null, diameter : Int): Drawable {
+    val icon = BitmapDrawable(context.resources, image.bitmap)
+            .apply {
+                iconColor?.let { setColorFilter(it, PorterDuff.Mode.SRC_IN) }
+            }
```
Comment (jiurchuk) [inline components/src/main/java/com/mercadolibre/android/andesui/utils/DrawableUtils.kt]: Creo que podriamos tener algo así como:
`val icon = buildColoredBitmapDrawable(...)` así reutilizamos la imple anterior.

File: components/src/main/java/com/mercadolibre/android/andesui/utils/DrawableUtils.kt [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+    val biggerCircle = ShapeDrawable(OvalShape())
+    val size = diameter
+    biggerCircle.intrinsicHeight = size
+    biggerCircle.intrinsicWidth = size
+    biggerCircle.paint.color = shapeColor!!
+    biggerCircle.bounds = Rect(0, 0, size, size)
+    return LayerDrawable(arrayOf(biggerCircle, icon ))
+    //FIXME Refactor this method
+
+
```
Comment (jiurchuk) [inline components/src/main/java/com/mercadolibre/android/andesui/utils/DrawableUtils.kt]: Empty lines

File: components/src/main/res/message/layout/andesui_layout_message.xml [hunk 8]
```diff
@@ -0,0 +1,101 @@
+<?xml version="1.0" encoding="utf-8"?>
+<android.support.v7.widget.CardView xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    app:cardCornerRadius="6dp"
```
Comment (jiurchuk) [inline components/src/main/res/message/layout/andesui_layout_message.xml]: Usemos `dimens`

File: components/src/main/res/message/layout/andesui_layout_message.xml [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+    <android.support.constraint.ConstraintLayout
+        android:id="@+id/andesui_message_container"
+        android:layout_width="match_parent"
+        android:layout_height="match_parent">
+
+        <TextView
+            android:id="@+id/andesui_title"
+            android:layout_width="0dp"
+            android:layout_height="wrap_content"
+            android:layout_marginBottom="8dp"
```
Comment (jiurchuk) [inline components/src/main/res/message/layout/andesui_layout_message.xml]: Dimens

File: components/build.gradle [hunk 10]
```diff
@@ -57,6 +57,7 @@ dependencies {
     //noinspection GradleDependency
```
Comment (eferrin) [inline components/build.gradle]: No deberia ser necesario. Revisar

File: components/src/main/java/com/mercadolibre/android/andesui/message/AndesMessage.kt [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+import com.mercadolibre.android.andesui.message.hierarchy.AndesMessageHierarchy
+import com.mercadolibre.android.andesui.message.state.AndesMessageState
+
+class AndesMessage : FrameLayout {
+
+    companion object {
+        private val HIERARCHY_DEFAULT = AndesMessageHierarchy.LOUD
+        private val STATE_DEFAULT = AndesMessageState.HIGHLIGHT
+        private val BODY_DEFAULT = "Andes Message Body"
+        private val TITLE_DEFAULT = null
```
Comment (eferrin) [inline components/src/main/java/com/mercadolibre/android/andesui/message/AndesMessage.kt]: Tiene sentido un Title Default?

File: components/src/main/java/com/mercadolibre/android/andesui/message/AndesMessage.kt (hunk 2 above)
Comment (eferrin) [reply components/src/main/java/com/mercadolibre/android/andesui/message/AndesMessage.kt]: No podes usar TextUtils.isEmpty()?

File: components/src/main/java/com/mercadolibre/android/andesui/message/AndesMessage.kt [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+    /**
+     * Gets data from the config and sets to the text component of this button.
+     *
+     */
+    private fun setupBodyComponent() {
+        bodyComponent.text = config.bodyText
+        bodyComponent.setTextSize(TypedValue.COMPLEX_UNIT_PX, config.bodySize)
+        bodyComponent.setTextColor(config.textColor)
+        bodyComponent.typeface = config.bodyTypeface
+//        bodyComponent.lineHeight = config.lineHeight //FIXME Use TextViewCompat
```
Comment (eferrin) [inline components/src/main/java/com/mercadolibre/android/andesui/message/AndesMessage.kt]: Comment? Se deberia borrar

File: components/src/main/java/com/mercadolibre/android/andesui/message/hierarchy/AndesMessageHierarchyInterface.kt [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+    /**
+     * Returns an [Int] representing a @ColorInt that will be used when tinting the icon.
+     *
+     * @param context needed for accessing some resources.
+     * @return an [Int] representing a @ColorInt that will be used when tinting the icon.
+     */
+    @ColorInt
+    abstract fun dismissableIconColor(context: Context): Int
+
+    //TODO Documentar
```
Comment (eferrin) [inline components/src/main/java/com/mercadolibre/android/andesui/message/hierarchy/AndesMessageHierarchyInterface.kt]: Documentar ajja

File: components/src/main/java/com/mercadolibre/android/andesui/message/state/AndesMessageStateInterface.kt [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+}
+
+internal object AndesHighlightMessageState : AndesMessageStateInterface() {
+    override fun primaryColor(context: Context) = ContextCompat.getColor(context, R.color.andesui_message_highlight_primary)
+    override fun secondaryColor(context: Context) = ContextCompat.getColor(context, R.color.andesui_message_highlight_primary_dark)
+    override fun icon(context: Context, hierarchy: AndesMessageHierarchyInterface) = buildColoredCircularShapeWithIconDrawable(
+            ContextCompat.getDrawable(context, R.drawable.andesui_ui_feedback_info_16) as BitmapDrawable,
+            context,
+            ContextCompat.getColor(context, R.color.andesui_white), hierarchy.iconBackgroundColor(context, this),
+            context.resources.getDimension(R.dimen.andesui_message_icon_diameter).toInt()
```
Comment (eferrin) [inline components/src/main/java/com/mercadolibre/android/andesui/message/state/AndesMessageStateInterface.kt]: Es necesario el toInt, no es un int ya

File: components/src/main/java/com/mercadolibre/android/andesui/utils/DrawableUtils.kt [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+                iconColor?.let { setColorFilter(it, PorterDuff.Mode.SRC_IN) }
+            }
+    val biggerCircle = ShapeDrawable(OvalShape())
+    val size = diameter
+    biggerCircle.intrinsicHeight = size
+    biggerCircle.intrinsicWidth = size
+    biggerCircle.paint.color = shapeColor!!
+    biggerCircle.bounds = Rect(0, 0, size, size)
+    return LayerDrawable(arrayOf(biggerCircle, icon ))
+    //FIXME Refactor this method
```
Comment (eferrin) [inline components/src/main/java/com/mercadolibre/android/andesui/utils/DrawableUtils.kt]: Delete Fixme?

File: demoapp/src/main/java/com/mercadolibre/android/andesui/demoapp/MessageShowcaseActivity.kt [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+        viewPager.adapter = AndesShowcasePagerAdapter(this)
+        val indicator = findViewById<PageIndicator>(R.id.page_indicator)
+        indicator.attach(viewPager)
+
+        val adapter = viewPager.adapter as AndesShowcasePagerAdapter
+        addMessage(adapter.views[0])
+    }
+
+    private fun addMessage(container: View) {
+        //TODO Create a Message view
```
Comment (eferrin) [inline demoapp/src/main/java/com/mercadolibre/android/andesui/demoapp/MessageShowcaseActivity.kt]: Esta bien este TODO?

File: demoapp/src/main/java/com/mercadolibre/android/andesui/demoapp/MessageShowcaseActivity.kt [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+
+        val linearLoud = container.findViewById<LinearLayout>(R.id.andes_loud_container)
+
+        //TODO Add messages to the linear layout
+        //linearLoud.addView(andesMessage)
```
Comment (eferrin) [inline demoapp/src/main/java/com/mercadolibre/android/andesui/demoapp/MessageShowcaseActivity.kt]: Estan bie  estos comments?

File: components/src/main/java/com/mercadolibre/android/andesui/message/AndesMessage.kt (hunk 12 above)
Comment (jiurchuk) [reply components/src/main/java/com/mercadolibre/android/andesui/message/AndesMessage.kt]: Lo dejamos a modo de reminder. En verdad deberíamos usar `TextViewCompat` que permite que seteemos el height de cada line. Sino es API 28

File: components/src/main/java/com/mercadolibre/android/andesui/message/state/AndesMessageStateInterface.kt (hunk 14 above)
Comment (jiurchuk) [reply components/src/main/java/com/mercadolibre/android/andesui/message/state/AndesMessageStateInterface.kt]: Nop, lamentablemente es necesario :(

File: demoapp/src/main/java/com/mercadolibre/android/andesui/demoapp/MessageShowcaseActivity.kt (hunk 16 above)
Comment (jiurchuk) [reply demoapp/src/main/java/com/mercadolibre/android/andesui/demoapp/MessageShowcaseActivity.kt]: Es parte de la test app, en la próxia iteración esto cobrará relevancia.

File: demoapp/src/main/java/com/mercadolibre/android/andesui/demoapp/MessageShowcaseActivity.kt (hunk 17 above)
Comment (jiurchuk) [reply demoapp/src/main/java/com/mercadolibre/android/andesui/demoapp/MessageShowcaseActivity.kt]: Same as https://github.com/mercadolibre/fury_andesui-android/pull/4#discussion_r370823475

**Pull request merged**